File: lib_common/src/main/java/com/fly/tour/common/manager/NewsDBManager.java
package com.fly.tour.common.manager;

import android.content.Context;

import com.fly.tour.common.R;
import com.fly.tour.common.util.log.KLog;
import com.fly.tour.db.dao.NewsDetailDao;
import com.fly.tour.db.dao.NewsTypeDao;
import com.fly.tour.db.entity.NewsDetail;
import com.fly.tour.db.entity.NewsType;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.lang.reflect.Type;
import java.util.List;

/**
 * Description: <NewsDBManager><br>
 * Author:      gxl<br>
 * Date:        2019/5/28<br>
 * Version:     V1.0.0<br>
 * Update:     <br>
 */
public class NewsDBManager {
    public static final String TAG = NewsDBManager.class.getSimpleName();
    private static NewsDBManager newsDBManager;
    private Context mContext;
    private List<NewsType> mListNewsType;

    private NewsDBManager(Context context) {
        mContext = context;
    }

    public static NewsDBManager getInstance(Context context) {
        if (newsDBManager == null) {
            synchronized (NewsDBManager.class) {
                if (newsDBManager == null) {
                    newsDBManager = new NewsDBManager(context);
                }
            }
        }
        return newsDBManager;
    }

    public void initNewsDB() {
        NewsTypeDao newsTypeDao = new NewsTypeDao(mContext);
        if (newsTypeDao.isEmpty()) {
            Gson gson = new Gson();
            Type type = new TypeToken<List<NewsType>>() {
            }.getType();
            mListNewsType = gson.fromJson(getStringByResId(R.raw.news_type), type);
            newsTypeDao.addListNewStype(mListNewsType);
        }
        NewsDetailDao newsDetailDao = new NewsDetailDao(mContext);
        if(newsDetailDao.isEmpty()){
            Gson gson = new Gson();
            Type type = new TypeToken<List<NewsDetail>>() {
            }.getType();
            List<NewsDetail> newsDetailList = gson.fromJson(getStringByResId(R.raw.news_detail), type);
            newsDetailDao.addListNewsDetail(newsDetailList);
        }
    }

    private String getStringByResId(int resid) {
        InputStream inputStream = mContext.getResources().openRawResource(resid);
        InputStreamReader inputStreamReader = new InputStreamReader(inputStream);
        BufferedReader bufferedReader = new BufferedReader(inputStreamReader);
        StringBuilder stringBuilder = new StringBuilder();
        String line = null;
        try {
            while ((line = bufferedReader.readLine()) != null) {
                stringBuilder.append(line);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return stringBuilder.toString();
    }

    public List<NewsType> getListNewsType() {
        return mListNewsType;
    }
}

File: lib_common/src/main/java/com/fly/tour/common/manager/NewsDBManager.kt
package com.fly.tour.common.manager

import android.content.Context

import com.fly.tour.common.R
import com.fly.tour.db.dao.NewsDetailDao
import com.fly.tour.db.dao.NewsTypeDao
import com.fly.tour.db.entity.NewsDetail
import com.fly.tour.db.entity.NewsType
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken

import java.io.BufferedReader
import java.io.File
import java.io.InputStreamReader

/**
 * Description: <NewsDBManager><br>
 * Author:      mxdl<br>
 * Date:        2019/5/28<br>
 * Version:     V1.0.0<br>
 * Update:     <br>
</NewsDBManager> */
class NewsDBManager private constructor(private val mContext: Context) {

    fun getListNewsType(): List<NewsType>? {
        return NewsTypeDao(mContext).getListNewsType()
    }

    fun initNewsDB() {
        val mNewsTypeDao = NewsTypeDao(mContext)
        if (mNewsTypeDao.isEmpty()) {
            val gson = Gson()
            val type = object : TypeToken<List<NewsType>>() {

            }.type
            val mListNewsType =
                gson.fromJson<List<NewsType>>(getStringByResId(R.raw.news_type), type)
            mNewsTypeDao.addListNewStype(mListNewsType)
        }
        val newsDetailDao = NewsDetailDao(mContext)
        if (newsDetailDao.isEmpty) {
            val gson = Gson()
            val type = object : TypeToken<List<NewsDetail>>() {

            }.type
            val newsDetailList =
                gson.fromJson<List<NewsDetail>>(getStringByResId(R.raw.news_detail), type)
            newsDetailDao.addListNewsDetail(newsDetailList)
        }
    }

    private fun getStringByResId(resid: Int): String {
        val inputStream = mContext.resources.openRawResource(resid)
        val inputStreamReader = InputStreamReader(inputStream)
        val bufferedReader = BufferedReader(inputStreamReader)
        val stringBuilder = StringBuilder()
        var line: String? = null
        try {
//            while (({line = bufferedReader.readLine()}) != null) {
//                stringBuilder.append(line)
//            }
            //以上为Java中的写法，在有些机型中会造成内存溢出，下面为Kotlin中的通用写法
            val text: List<String> = bufferedReader.readLines()
            for (line in text) {
                stringBuilder.append(line)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            bufferedReader.close()
            inputStreamReader.close()
            inputStream.close()
        }
        return stringBuilder.toString()
    }

    companion object {
        val TAG = NewsDBManager::class.java!!.getSimpleName()
        private var newsDBManager: NewsDBManager? = null

        fun getInstance(context: Context): NewsDBManager? {
            if (newsDBManager == null) {
                synchronized(NewsDBManager::class.java) {
                    if (newsDBManager == null) {
                        newsDBManager = NewsDBManager(context)
                    }
                }
            }
            return newsDBManager
        }
    }
}
